From e0ffea7da45a3dc36d3957022672ce6890fda214 Mon Sep 17 00:00:00 2001
From: bebbo <bebbo@bejy.net>
Date: Sun, 25 Jan 2026 12:54:36 +0100
Subject: [PATCH 01/15] fix index out of bounds, if no insn after a label

---
 gcc/bbb-opts.c | 15 ++++++++++-----
 1 file changed, 10 insertions(+), 5 deletions(-)

diff --git a/gcc/bbb-opts.c b/gcc/bbb-opts.c
index f0549d86df7..717b6c82efb 100644
--- a/gcc/bbb-opts.c
+++ b/gcc/bbb-opts.c
@@ -2907,12 +2907,17 @@ opt_reg_rename (void)
 		      int labi = 0;
 
 		      // search next non-label insn
-		      insn_info * bb;
-		      do
-		        bb = &(*infos)[label_index + ++labi];
-		      while (bb->is_label());
+		      insn_info *bb = NULL;
+		      while (label_index + labi + 1 < infos->size())  // ensure we stay within bounds
+			{
+			  labi++;
+			  bb = &(*infos)[label_index + labi];
+			  if (!bb->is_label ())
+			    break;
+			  bb = NULL;
+			}
 
-		      if (found.find (label_index) == found.end () && bb->is_use (rename_regno))
+		      if (bb != NULL && found.find (label_index) == found.end () && bb->is_use (rename_regno))
 			{
 //			  printf ("jump %d -> label %d \n", pos, label_index); fflush (stdout);
 			  todo.insert (label_index);
