diff --git a/Makefile b/Makefile
index 1c69d24..4b43f3d 100644
--- a/Makefile
+++ b/Makefile
@@ -1197,14 +1197,17 @@ branch:
 # =================================================
 # multilib support
 # =================================================
-MULTI = MODNAME/.: \
-		MODNAME/libm020:-m68020 \
-		MODNAME/libm020/libm881:-m68020_-m68881 \
-		MODNAME/libb:-fbaserel \
-		MODNAME/libb/libm020:-fbaserel_-m68020 \
-		MODNAME/libb/libm020/libm881:-fbaserel_-m68020_-m68881 \
-		MODNAME/libb32/libm020:-fbaserel32_-m68020 \
-		MODNAME/libb32/libm020/libm881:-fbaserel32_-m68020_-m68881
+MULTI = MODNAME/.: \
+		MODNAME/libm060:-m68060 \
+		MODNAME/libm020:-m68020 \
+		MODNAME/libm020/libm881:-m68020_-m68881 \
+		MODNAME/libb:-fbaserel \
+		MODNAME/libb/libm060:-fbaserel_-m68060 \
+		MODNAME/libb/libm020:-fbaserel_-m68020 \
+		MODNAME/libb/libm020/libm881:-fbaserel_-m68020_-m68881 \
+		MODNAME/libb32/libm060:-fbaserel32_-m68060 \
+		MODNAME/libb32/libm020:-fbaserel32_-m68020 \
+		MODNAME/libb32/libm020/libm881:-fbaserel32_-m68020_-m68881
-
+
 # 1=module name, 2=from name, 3 = to name
 COPY_MULTILIBS = $(foreach T, $(subst MODNAME,$1,$(MULTI)),cp $(BUILD)/$(word 1,$(subst :, ,${T}))/$2 $(BUILD)/$(word 1,$(subst :, ,${T}))/$3;)
