diff --git a/src/Makefile b/src/Makefile
index 2c2f328..48e4659 100644
--- a/src/Makefile
+++ b/src/Makefile
@@ -500 +500 @@ $(TARGET): $(CATFILES) $(addprefix $(OBJDIR)/,$(OBJS))
-	@$(CC) $(LDFLAGS) -o $@.debug $(addprefix $(OBJDIR)/,$(OBJS)) $(LDLIBS) -Wl,--cref,-M,-Map=$@.map
+	@$(CC) $(LDFLAGS) -o $@.debug $(addprefix $(OBJDIR)/,$(OBJS)) $(LDLIBS)
@@ -502 +502 @@ $(TARGET): $(CATFILES) $(addprefix $(OBJDIR)/,$(OBJS))
-	@$(STRIP) --preserve-dates -R.comment -R.sdata2 -S -o $@ $@.debug
+	@$(CP) $@.debug $@
diff --git a/src/version.h b/src/version.h
index 382b08b..2d7a68a 100644
--- a/src/version.h
+++ b/src/version.h
@@ -79,0 +80,4 @@
+#elif defined(__aarch64__) || defined(__arm64__)
+  #define CPU "arm64"
+#elif defined(__arm__)
+  #define CPU "arm"
